python-stdlib/unittest/unittest.py

```diff
@@ -148,82 +236,283 @@ def skipUnless(cond, msg):
+        for c, e in lst:
+            detail = " ".join((str(i) for i in c))
+            print("======================================================================")
+            print(f"FAIL: {detail}")
+            print(sep)
+            print(e)
+
+    def __repr__(self):
+        # Format is compatible with CPython.
+        return "<unittest.result.TestResult run=%d errors=%d failures=%d>" % (
+            self.testsRun,
+            self.errorsNum,
+            self.failuresNum,
+        )
+
+    def __add__(self, other):
+        self.errorsNum += other.errorsNum
+        self.failuresNum += other.failuresNum
+        self.skippedNum += other.skippedNum
+        self.testsRun += other.testsRun
+        self.errors.extend(other.errors)
+        self.failures.extend(other.failures)
+        self.skipped.extend(other.skipped)
+        return self
 
-# TODO: Uncompliant
-def run_class(c, test_result):
-    o = c()
+
+def capture_exc(exc, traceback):
+    buf = io.StringIO()
+    if hasattr(sys, "print_exception"):
+        sys.print_exception(exc, buf)
+    elif traceback is not None:
+        traceback.print_exception(None, exc, traceback, file=buf)
+    return buf.getvalue()
+
+
+def handle_test_exception(
+    current_test: tuple, test_result: TestResult, exc_info: tuple, verbose=True
+):
+    exc = exc_info[1]
+    traceback = exc_info[2]
+    ex_str = capture_exc(exc, traceback)
+    if isinstance(exc, AssertionError):
+        test_result.failuresNum += 1
+        test_result.failures.append((current_test, ex_str))
+        if verbose:
+            print(" FAIL")
+    else:
+        test_result.errorsNum += 1
+        test_result.errors.append((current_test, ex_str))
+        if verbose:
+            print(" ERROR")
+    test_result._newFailures += 1
+
+
+def run_suite(c, test_result: TestResult, suite_name=""):
+    if isinstance(c, TestSuite):
+        c.run(test_result)
+        return
+
+    if isinstance(c, type):
+        o = c()
+    else:
+        o = c
+    set_up_class = getattr(o, "setUpClass", lambda: None)
+    tear_down_class = getattr(o, "tearDownClass", lambda: None)
     set_up = getattr(o, "setUp", lambda: None)
     tear_down = getattr(o, "tearDown", lambda: None)
+    exceptions = []
+    try:
+        suite_name += "." + c.__qualname__
+    except AttributeError:
+        pass
+
+    def run_one(test_function):
+        global __test_result__, __current_test__
+        print("%s (%s) ..." % (name, suite_name), end="")
+        set_up()
+        __test_result__ = test_result
+        test_container = f"({suite_name})"
+        __current_test__ = (name, test_container)
+        try:
+            test_result._newFailures = 0
+            test_result.testsRun += 1
+            test_globals = dict(**globals())
+            test_globals["test_function"] = test_function
+            exec("test_function()", test_globals, test_globals)
```

Comment:
Ah, this was all done so that each test is run in a clean python environment, avoiding tests unexpectedly breaking because a previous test has modified the global environment in some way. This was particularly helpful with big test suits running lots of tests were it can be very hard to track down an earlier test that hasn't quite reverted any changes made... Or conversely tests passing only because a previous one has inadvertently set/overridden something.

This was done to make tests simpler and more reliable... however in hindsight I never thought to check for explicit cpython compatibility on this point.

Reply:
Is `test_NotChangedByOtherTest` supposed to test this? Because that tests succeeeds here, both in python and micropython, and in the latter also after removing the `test_globals` code. I'm not completely sure why, but it's not quite clear what it is actually being tested now.

> This was all done so that each test is run in a clean python environment, avoiding tests unexpectedly breaking because a previous test has modified the global environment in some way

Arguably those tests should use setup/teardown correctly; but in any case: 

> globals()
>  Return a dictionary representing the current global symbol table. This is always the dictionary of the current module (inside a function or method, this is the module where it is defined, not the module from which it is called).

so I wonder whether this actually changed anything for you, i.e. this is just the globals for the unittest module, which doesn't affect tests which modify their own global environment or globals in another module? Or am I missing something here?

Reply:
I've re-visited this code myself.... and yeah I think it's a bit broken. My "test isolation" feature is mostly based on caching / resetting `sys.modules` to force a reload of all the imports clean for each test. This was intended to blow away any monkey patching etc. that might have been done and not fixed by previous tests.

I think this `exec()` line was left over from an earlier iteration of this feature and really doen't add value here.

You're entirely correct that setup/teardown functions should be used to reverse anything like this on tests, I was basically trying to protect developers (eg me) from themselves by reducing this need... on many projects in the past I and others have lost a lot of time chasing down intermittent test failures that can be due to really subtle changes in timing / order of tests that this could certainly help with!

On a related note, the time when `test_NotChangedByOtherTest` failed for me yesterday is when running it via `micropython ./test_unittest.py` rather than `micropython -m unittest`. 

When `./test_unittest.py` is the main test, its imports / global values are the initial values that get cached in `_snapshot_modules()` function so will be made available to each test in `run_module()` when `discover` is later run. Certainly `discover` was not supposed to be run here though...

Reply:
Ok so the test_globals stuff can go. What about _snapshot_modules (also not CPython-compatible) and test_unittest_isolated (doesn't actually test anuthing iunittest-specific)?

Reply:
I'm conflicted about the `_snapshot_modules` stuff because while it's not entirely consistent with cpython, I find it very useful. I wonder if there's a way I could pull that out into a separate module / hook / etc.

Reply:
Thing is, sometimes it might be wanted, exactly to be able to verify inter-modle dependencies (whether that's good is something else). So the only sane thing would be an argument?

Reply:
The `_snapshot_modules` resets the state between separate `test*.py` files, not between individual functions within a test file. It would be a very fragile test suite if you wanted state from one test file to be configured / shared to other test files... surely?

Reply:
But yeah, considering it's non-standard behavior maybe it's better as an extension TestRunner that can be called something more recognisable for people to use.